tools/model_tools/src/openvino/model_zoo/download_engine/downloader.py

```diff
@@ -12,24 +12,71 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import contextlib
 import functools
 import requests
 import ssl
+import sys
+import threading
 import time
 import types
 
+from pathlib import Path
+from typing import Set
+
+from openvino.model_zoo import _common, _concurrency, _reporting
 from openvino.model_zoo.download_engine import cache
 
 DOWNLOAD_TIMEOUT = 5 * 60
 
+
+# There is no evidence that the requests.Session class is thread-safe,
+# so for safety, we use one Session per thread. This class ensures that
+# each thread gets its own Session.
+class ThreadSessionFactory:
+    def __init__(self, exit_stack):
+        self._lock = threading.Lock()
+        self._thread_local = threading.local()
+        self._exit_stack = exit_stack
+
+    def __call__(self):
+        try:
+            session = self._thread_local.session
+        except AttributeError:
+            with self._lock: # ExitStack might not be thread-safe either
+                session = self._exit_stack.enter_context(requests.Session())
+            self._thread_local.session = session
+        return session
+
+
 class Downloader:
-    def __init__(self, output_dir=None, cache_dir=None, num_attempts=1, timeout=DOWNLOAD_TIMEOUT):
+    def __init__(self, requested_precisions: str = None, output_dir: Path = None,
+                 cache_dir: Path = None, num_attempts: int = 1, timeout: int = DOWNLOAD_TIMEOUT):
         self.output_dir = output_dir
         self.cache = cache.NullCache() if cache_dir is None else cache.DirCache(cache_dir)
         self.num_attempts = num_attempts
         self.timeout = timeout
+        self.requested_precisions = requested_precisions
+
+    @property
+    def requested_precisions(self) -> Set[str]:
+        return self._requested_precisions
+
+    @requested_precisions.setter
+    def requested_precisions(self, value: str = None):
+        if value is None:
+            _requested_precisions = _common.KNOWN_PRECISIONS
+        else:
+            _requested_precisions = set(value.split(','))
+
+        unknown_precisions = _requested_precisions - _common.KNOWN_PRECISIONS
+        if unknown_precisions:
+            sys.exit('Unknown precisions specified: {}.'.format(', '.join(sorted(unknown_precisions))))
+
+        self._requested_precisions = _requested_precisions
 
     def _process_download(self, reporter, chunk_iterable, size, progress, file):
+        #same
```

Comment:
very informative comment :) is it really required or it is just debug info?

Reply:
No, just skipped it accidentally. Thanks for noticing.